File: 135-Candy/Candy.go
package candy

func candy(ratings []int) int {
	n := len(ratings)

	toLeft := make([]int, n)
	toRight := make([]int, n)

	toLeft[n-1] = 1
	toRight[0] = 1

	for i := 1; i < n; i++ {
		if ratings[i] > ratings[i-1] {
			toRight[i] = toRight[i-1] + 1
		} else {
			toRight[i] = 1
		}
	}

	for i := n - 2; i >= 0; i-- {
		if ratings[i] > ratings[i+1] {
			toLeft[i] = toLeft[i+1] + 1
		} else {
			toLeft[i] = 1
		}
	}

	count := 0
	for i := 0; i < n; i++ {
		if toLeft[i] > toRight[i] {
			count += toLeft[i]
		} else {
			count += toRight[i]
		}
	}

	return count
}

File: 0135-Candy/Candy.java

class Candy {
    class Solution {
        public int candy(int[] ratings) {
            int[] toLeft = new int[ratings.length];
            int[] toRight = new int[ratings.length];
            toLeft[ratings.length - 1] = 1;
            toRight[0] = 1;
            
            
            for(int i = 1; i < ratings.length; i++){
                toRight[i] = ratings[i] > ratings[i-1] ? toRight[i - 1] + 1 : 1;
            }
            
            for(int i = ratings.length - 2; i >= 0; i--){
                toLeft[i] = ratings[i] > ratings[i + 1] ? toLeft[i + 1] + 1 : 1;
            }
            
            int count = 0;
            for(int i = 0; i < ratings.length; i++){
                count += Math.max(toLeft[i], toRight[i]);
            }
            return count;
        }
    }
    
}
